### PR TITLE
Run GRPC server over TLS

### PR DESCRIPTION
**Pull request by lookuptable**
**Release note**:

```
GRPC service is now served over TLS
```

### CODE REVIEW

File: pkg/pki/ca/ca.go [hunk 1]
```diff
@@ -137,8 +137,14 @@ func (ca *IstioCA) GetRootCertificate() []byte {
...
 }
 
-// Sign signs a CSR
-func (ca *IstioCA) Sign(csr *x509.CertificateRequest) ([]byte, error) {
+// Sign takes a PEM-encoded certificate signing request and returns a signed
+// certificate.
+func (ca *IstioCA) Sign(csrPEM []byte) ([]byte, error) {
+	csr, err := pki.ParsePemEncodedCSR(csrPEM)
+	if err != nil {
+		return nil, err
```
Comment (wattli) [inline pkg/pki/ca/ca.go]: Add some logging info for debugging purpose?

File: pkg/server/grpc/server.go [hunk 2]
```diff
@@ -88,6 +88,59 @@ func (s *Server) Run() error {
...
+		hostname: hostname,
+		port:     port,
+	}
+}
+
+func (s *Server) createTLSServerOption() grpc.ServerOption {
+	config := &tls.Config{
+		GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			if s.certificate == nil || !isValid(s.certificate) {
+				// apply new certificate if there isn't one yet, or the one has become invalid
```
Comment (wattli) [inline pkg/server/grpc/server.go]: s/apply/Apply

File: pkg/server/grpc/server.go [hunk 3]
```diff
@@ -88,6 +88,59 @@ func (s *Server) Run() error {
...
+	return grpc.Creds(credentials.NewTLS(config))
+}
+
+func (s *Server) applyServerCertificate() (*tls.Certificate, error) {
+	opts := ca.CertOptions{
+		Host:       s.hostname,
+		RSAKeySize: 1024,
+	}
+
+	csrPEM, privPEM, err := ca.GenCSR(opts)
```
Comment (wattli) [inline pkg/server/grpc/server.go]: Is this the self-signed cert option for CA?

File: pkg/server/grpc/server.go [hunk 4]
```diff
@@ -88,6 +88,59 @@ func (s *Server) Run() error {
...
+				}
+				s.certificate = newCert
+			}
+			return s.certificate, nil
+		},
+	}
+	return grpc.Creds(credentials.NewTLS(config))
+}
+
+func (s *Server) applyServerCertificate() (*tls.Certificate, error) {
```
Comment (wattli) [inline pkg/server/grpc/server.go]: I think the CA should get key/cert from some file via volume mount, no?

File: pkg/pki/ca/ca.go (hunk 1 above)
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: It'll logged by the caller.

File: pkg/server/grpc/server.go (hunk 3 above)
Comment (lookuptable) [inline pkg/server/grpc/server.go]: It's rooted at the same CA that is used for signing workloads, regardless that CA is self-signed or not.

File: pkg/server/grpc/server.go [hunk 5]
```diff
@@ -88,6 +88,59 @@ func (s *Server) Run() error {
...
+	return &cert, nil
+}
+
+func isValid(cert *tls.Certificate) bool {
+	leaf := cert.Leaf
+	if leaf == nil {
+		return false
+	}
+	now := time.Now()
+	return leaf.NotBefore.Before(now) && leaf.NotAfter.After(now)
```
Comment (lizan) [inline pkg/server/grpc/server.go]: Leave some buffer time here, to make sure when client receives the certificate it is still valid. (30secs or 60secs should do it).

File: pkg/server/grpc/server.go [hunk 6]
```diff
@@ -88,6 +88,59 @@ func (s *Server) Run() error {
...
+			return s.certificate, nil
+		},
+	}
+	return grpc.Creds(credentials.NewTLS(config))
+}
+
+func (s *Server) applyServerCertificate() (*tls.Certificate, error) {
+	opts := ca.CertOptions{
+		Host:       s.hostname,
+		RSAKeySize: 1024,
```
Comment (lizan) [inline pkg/server/grpc/server.go]: Thought 1024-bit RSA is not considered secure enough [(ref)](https://knowledge.rapidssl.com/support/ssl-certificate-support/index?page=content&id=INFO1684&actp=search&viewlocale=en_US&searchid=1442967219430), use 2048?

**Comment by lookuptable** [comment]
@wattli @lizan PTAL

File: pkg/server/grpc/server.go [hunk 7]
```diff
@@ -88,6 +90,60 @@ func (s *Server) Run() error {
...
+	return &cert, nil
+}
+
+func isValid(cert *tls.Certificate) bool {
+	leaf := cert.Leaf
+	if leaf == nil {
+		return false
+	}
+	now := time.Now()
+	return leaf.NotBefore.Add(certValidationBuffer).Before(now) &&
```
Comment (lizan) [inline pkg/server/grpc/server.go]: Do we need the buffer for `NotBefore`? Won't this cause a certificate to be considered invalid right after issuance?

Comment (lookuptable) [inline pkg/server/grpc/server.go]: I see.

Thinking a little more, I guess it makes sense to rename this method to `needToRefreshCert()`, and skip `NotBefore` check totally.

**Pull request merged**